File: lib/ellipse_functions.cc

#include <iostream>
#include <math.h>

#include "ellipse_functions.h"




std::vector<double> polar_to_cartesian(double r, double theta){

	std::vector<double> coords {r*cos(theta), r*sin(theta)};
	return coords;

}

std::vector<double> cartesian_to_polar(double x, double y){
	
	std::vector<double> coords {sqrt(x*x + y*y), atan(y/x)};
	return coords;
}

double eccentricity (double a, double b){ return sqrt(1-pow(a/b, 2));}

double r(double theta, double a, double b, double alpha){
	return a*b/sqrt(pow((b*cos(theta-alpha)),2) + pow((a*sin(theta-alpha)),2));
}


void ellipse_cut ( std::vector<std::vector<double> >* res, std::vector<std::vector<double> >* p, std::vector<double> lower_ellipse, std::vector<double> upper_ellipse){

	double a_lower, b_lower, alpha_lower, a_upper, b_upper, alpha_upper;
	std::vector<std::vector<double> > p_new;
	
	a_lower = lower_ellipse[0] ; b_lower = lower_ellipse[1] ; alpha_lower=0;
	a_upper = upper_ellipse[0] ; b_upper = upper_ellipse[1] ; alpha_upper=0;
	
	if (lower_ellipse.size() == 3){alpha_lower = lower_ellipse[2];}
	if (upper_ellipse.size() == 3){alpha_upper = upper_ellipse[2];}

	double radius, lower_radius, upper_radius, theta;
	for ( auto coord : *p ){
		radius = sqrt(pow(coord[0], 2) + pow(coord[1], 2));
		
		if (coord[1] != 0){theta = atan(coord[1]/coord[0]);}
		else if (coord[0] == 0){theta = M_PI/2;}
		else {theta = 0;}
	
		lower_radius = r(theta, a_lower, b_lower, alpha_lower);
		upper_radius = r(theta, a_upper, b_upper, alpha_upper);

		if (lower_radius < radius && upper_radius > radius){
				p_new.push_back(coord);
		}
	}

	*res = p_new;

}

File: tests/cds_data_and_plots/ellipse_functions.py
import numpy as np
from numpy import cos, sin

FULL_THETA = np.array([2*np.pi*(x/100) for x in range(0,101)])

# returns (x, y) coordinates given (r, theta)
def polar_to_cartesian(r, theta):
    x = r*cos(theta)
    y = r*sin(theta)
    return x, y

# returns the eccentricity of an ellipse with 
# (semi-major, semiminor) = (a, b)
def eccentricity(a, b):
    return np.sqrt(1-(b/a)**2)

# returns the distance from the origin for a 
# point on an ellipse and angle theta from 
# x-axis with (semi-major, semiminor) = (a, b)
def r(theta, a, b, alpha=0):
    return a*b/np.sqrt((b*np.cos(theta-alpha))**2 + (a*np.sin(theta-alpha))**2)

# cuts out all stars within lower_ellipse (with
# semi-major axis a) and all stars outside of 
# the ellipse that has semi-major axis 
# a' = a + diff but the SAME eccentricity
def basic_ellipse_cut(p, lower_ellipse, diff):

    a_lower = lower_ellipse[0] ; b_lower = lower_ellipse[1]
    a_upper = a_lower+diff ; b_upper = a_upper*(b_lower/a_lower)

    return ellipse_cut(p, (a_lower, b_lower), (a_upper, b_upper))


# cuts out all stars within lower_ellipse (with
# semi-major axis a) and all stars outside of 
# the ellipse that has semi-major axis 
# a' = a + diff but DIFFERENT eccentricity
def ellipse_cut(p, lower_ellipse, upper_ellipse, center=(0,0)):

    a_lower = lower_ellipse[0] ; b_lower = lower_ellipse[1] ; alpha_lower = 0
    a_upper = upper_ellipse[0] ; b_upper = upper_ellipse[1] ; alpha_upper = 0

    if len(lower_ellipse) == 3:
        alpha_lower = lower_ellipse[2]
    if len(upper_ellipse) == 3:
        alpha_upper = upper_ellipse[2]

    cx, cy = center
    x = p[:,0]+cx ; y = p[:,1]+cy
    rl = [np.sqrt(x**2 + y**2) for x, y in zip(x, y)]
    theta = [np.arctan2(y,x) for x, y in zip(x, y)]

    inner_cut = [ ri > r(theta, a_lower, b_lower, alpha=alpha_lower) for ri, theta in zip(rl, theta)]
    outer_cut = [ ri < r(theta, a_upper, b_upper, alpha=alpha_upper) for ri, theta in zip(rl, theta)]

    cut = [i and o for i,o in zip(inner_cut, outer_cut)]

    return p[cut]

